### PR TITLE
nixos/systemd: remove support for legacy cgroup hierachy

### PR DESCRIPTION
**Pull request by arianvp**
## Description of changes

the `systemd.enableUnifiedCgroupHierarchy` option has been removed. In systemd 256 support for cgroup v1 ('legacy' and 'hybrid' hierarchies) is now considered obsolete and systemd by default will refuse to boot under it.
  To forcibly reenable cgroup v1 support, you can `set boot.kernelParams = [ "systemd.unified_cgroup_hierachy=0" "SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1" ]`.  NixOS does not officially support this configuration and might cause your system to be unbootable in future versions. You are on your own.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by arianvp** [comment]
cc @Artturin as this breaks some behaviour of the ananicy module and I can't really fix it? https://gitlab.com/ananicy-cpp/ananicy-cpp/-/issues/40

**Comment by arianvp** [comment]
@GrahamcOfBorg test lxd

**Comment by arianvp** [comment]
@GrahamcOfBorg test lxd

**Comment by flokli** [comment]
> cc @Artturin as this breaks some behaviour of the ananicy module and I can't really fix it? https://gitlab.com/ananicy-cpp/ananicy-cpp/-/issues/40

cgroupsv1 are dead, and have been for years. If this project didn't react on it, it'll probably stop working alltogether at some point, if they don't act. You can try the workaround provided above, but you're on your own, and probably should look for alternatives or fix it yourself.

File: nixos/doc/manual/release-notes/rl-2411.section.md [hunk 1]
```diff
@@ -96,6 +96,14 @@
 
 - `nginx` package no longer includes `gd` and `geoip` dependencies. For enabling it, override `nginx` package with the optionals `withImageFilter` and `withGeoIP`.
 
+- the `systemd.enableUnifiedCgroupHierarchy` option has been removed.
+  In systemd 256 support for cgroup v1 ('legacy' and 'hybrid' hierarchies) is
+  now considered obsolete and systemd by default will refuse to boot under it.
+  To forcibly reenable cgroup v1 support, you can `set boot.kernelParams = [
+  "systemd.unified_cgroup_hierachy=0" "SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1" ]`.
+  NixOS does not officially support this configuration and might cause your
+  system to be unbootable in future versions. You are on your own.
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2411.section.md:105]:
```suggestion
- The `systemd.enableUnifiedCgroupHierarchy` option has been removed.
  In systemd 256 support for cgroup v1 ('legacy' and 'hybrid' hierarchies) is
  now considered obsolete and systemd by default will refuse to boot under it.
  To forcibly reenable cgroup v1 support, you can
  `set boot.kernelParams = [ "systemd.unified_cgroup_hierachy=0" "SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1" ]`.
  NixOS does not officially support this configuration and might cause your
  system to be unbootable in future versions. You are on your own.
```
lets make it easy to grep for that message

File: nixos/modules/services/misc/ananicy.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/ananicy.nix:1]: Do we completely break it or only parts? If it would be the former we should add an assertion IMO

Comment (flokli) [reply nixos/modules/services/misc/ananicy.nix:1]: If this doesn't work with any supported kernel configuration, the module should be removed IMHO.

Comment (arianvp) [reply nixos/modules/services/misc/ananicy.nix:1]: The linked issue says that some things break.  

I don't know what you mean with an assertion in this case.

File: nixos/modules/services/misc/ananicy.nix [hunk 2]
```diff
@@ -116,7 +116,6 @@ in
       } // (if ((lib.getName cfg.package) == (lib.getName pkgs.ananicy-cpp)) then {
         # https://gitlab.com/ananicy-cpp/ananicy-cpp/-/blob/master/src/config.cpp#L12
         loglevel = mkOD "warn"; # default is info but its spammy
-        cgroup_realtime_workaround = mkOD config.systemd.enableUnifiedCgroupHierarchy;
```
Comment (K900) [inline nixos/modules/services/misc/ananicy.nix:119]: This should probably always be true now?

Comment (arianvp) [reply nixos/modules/services/misc/ananicy.nix:119]: The default is true: https://gitlab.com/ananicy-cpp/ananicy-cpp/-/blob/master/src/config.cpp#L22

Comment (arianvp) [reply nixos/modules/services/misc/ananicy.nix:119]: but added it for clarity now

**Review by ElvishJerricco — APPROVED** [review]

**Review by flokli — APPROVED** [review]

**Pull request merged**